Move SM2 to its own PMETH

randombit opened the pull request:
This moves SM2 to its own PMETH following the approach suggested by @mattcaswell in
https://github.com/openssl/openssl/pull/6402 fixing issue #5924

This is a contribution from Ribose Inc.

CC @ronaldtse

mattcaswell commented on crypto/evp/p_lib.c:
```diff
@@ -356,6 +356,41 @@ int EVP_PKEY_set_type_str(EVP_PKEY *pkey, const char *str, int len)
+        return 0;
+    }
+
+    pkey->save_type = pkey->type;
```
I think `save_type` is the original type of the object, so should probably be left unchanged.

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
```
Why do we need this?

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+static int pkey_sm2_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
+{
+    SM2_PKEY_CTX *dctx, *sctx;
+    if (!pkey_sm2_init(dst))
```
Insert a blank line between declarations and this line

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        return 0;
+    sctx = src->data;
+    dctx = dst->data;
+    if (sctx->gen_group) {
```
Style: `sctx->gen_group != NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    dctx = dst->data;
+    if (sctx->gen_group) {
+        dctx->gen_group = EC_GROUP_dup(sctx->gen_group);
+        if (!dctx->gen_group)
```
`dctx->gen_group == NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+static void pkey_sm2_cleanup(EVP_PKEY_CTX *ctx)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    if (dctx) {
```
Blank line above. Also `dctx != NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    if (!sig) {
```
`sig == NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        return 0;
+    }
+
+    if (dctx->md)
```
`dctx->md != NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    else
+        type = NID_sm3;
+
+    ret = sm2_sign(type, tbs, tbslen, sig, &sltmp, ec);
```
I'm wondering why the digest len and sig len parameters are declared as `unsigned int` instead of `size_t` in the `sm2_sign` function. Wouldn't `size_t` be more appropriate? Probably that's for a different PR though.

randombit replied:
I am personally a fan of using size_t for anything that is a length. I think I used `int` for digest len and `unsigned int` for output parameter to match `ECDSA_sign`. That avoided having to do type conversions, when ECDSA+SM2 shared a PMETH.

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    if (dctx->md)
```
`dctx->md != NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    int ret;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    int md_type;
```
No blank line in declarations

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    int md_type;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx->md)
```
`dctx->md != NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        else
+            ret = 1;
+    }
+    else {
```
`} else {`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    int ret = -1;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    int md_type;
```
No blank line above

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    int md_type;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx->md)
```
`dctx->md "= NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        else
+            ret = 1;
+    }
+    else {
```
`} else {`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_GROUP *group;
+    switch (type) {
```
Blank line after declarations

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        return 1;
+
+    case EVP_PKEY_CTRL_EC_PARAM_ENC:
+        if (!dctx->gen_group) {
```
`dctx->gen_group == NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+{
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+        int nid;
+        nid = EC_curve_nist2nid(value);
```
Blank line after declarations

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        return EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, nid);
+    } else if (strcmp(type, "ec_param_enc") == 0) {
+        int param_enc;
+        if (strcmp(value, "explicit") == 0)
```
blank line after declarations

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    EC_KEY *ec = NULL;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    int ret = 0;
+    if (dctx->gen_group == NULL) {
```
Blank line after declarations

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+    if (ec == NULL)
+        return 0;
+    ret = EC_KEY_set_group(ec, dctx->gen_group);
+    if (ret)
```
`ret != NULL`

randombit replied:
`ret` is an int not pointer. Do you want `ret != 0` here?

mattcaswell replied:
Oops. My mistake. Leave it as it is.

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+{
+    EC_KEY *ec = NULL;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    if (ctx->pkey == NULL && dctx->gen_group == NULL) {
```
Blank line after declarations

stevelinsell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,322 @@
+        return 0;
+    }
+    ec = EC_KEY_new();
+    if (!ec)
```
Consistency. At line 256 you do the equivalent check of ec == NULL but in this function you do !ec. Change the !ec to ec == NULL as well.

randombit commented:
@mattcaswell All review comments addressed. I also added a basic doc. I didn't find an existing pod that seemed to make sense, so I created `doc/man3/EVP_PKEY_type.md` to document all the various (currently undocumented) functions for setting/getting EVP_PKEY type. However beyond giving the declarations, only `EVP_PKEY_set_alias_type` is actually documented there.

mattcaswell commented on crypto/evp/p_lib.c:
```diff
@@ -356,6 +356,40 @@ int EVP_PKEY_set_type_str(EVP_PKEY *pkey, const char *str, int len)
+    if (ameth == NULL) {
+        EVPerr(EVP_F_EVP_PKEY_SET_ALIAS_TYPE, EVP_R_UNSUPPORTED_ALGORITHM);
+        return 0;
+    }
```
I think this ameth stuff can probably be removed now that you amended the "if" condition below to not need it.

dot-asm replied:
It stung my eye too. I mean it is redundant, because EVP_PKEY_type below calls EVP_PKEY_asn1_find and check for ameth being non-NULL.

And while you're at it, style guide suggests that block comments get stars aligned at opening star, i.e. and 4*n+1 position, not 4*n.

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+        if (EVP_MD_type((const EVP_MD *)p2) != NID_sha256 &&
+            EVP_MD_type((const EVP_MD *)p2) != NID_sha384 &&
+            EVP_MD_type((const EVP_MD *)p2) != NID_sha512 &&
+            EVP_MD_type((const EVP_MD *)p2) != NID_sm3) {
```
Sorry missed this on the first read through. Please put the "&&" on the beginning of the next line.

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    if (ec == NULL)
+        return 0;
+    EVP_PKEY_assign_EC_KEY(pkey, ec);
+    if (ctx->pkey) {
```
`ctx->pkey != NULL`

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+        if (!EVP_PKEY_copy_parameters(pkey, ctx->pkey))
+            return 0;
+    } else {
+        if (!EC_KEY_set_group(ec, dctx->gen_group))
```
We don't seem to then use `ec` for anything after this point.

randombit replied:
I am not sure what I should change. This function is copied from same function in `ec_pmeth.c` fwiw

mattcaswell replied:
Forget it - my mistake. It's assigned to the pkey above here.

mattcaswell commented on doc/man3/EVP_PKEY_type.md:
```diff
@@ -0,0 +1,43 @@
+EVP_PKEY_base_id,
+EVP_PKEY_set_type,
+EVP_PKEY_set_type_str,
+EVP_PKEY_set_alias_type,
```
I think we should either document all of these, or not include them in the pod at all (and remove them from the synopsis below). If you go with the latter of those two options then rename the file to match the one function that you are documenting. The problem is if we list these undocumented functions here then our scripts for finding undocumented functions will stop reporting them!

randombit replied:
OK

randombit replied:
Turns out EVP_PKEY_type, EVP_PKEY_id and EVP_PKEY_base_id were already documented, just not where I expected (EVP_PKEY_set1_RSA). I will add the documentation of EVP_PKEY_set_alias_type to this same doc so at least they are all together.

mattcaswell commented on doc/man3/EVP_PKEY_type.md:
```diff
@@ -0,0 +1,43 @@
+EVP_PKEY_set_type,
+EVP_PKEY_set_type_str,
+EVP_PKEY_set_alias_type,
+- examine and manipulate type of public key
```
Perhaps: "examine and manipulate the type of a key"? EVP_PKEY could be a private/public key pair or just a public key.

mattcaswell commented on doc/man3/EVP_PKEY_type.md:
```diff
@@ -0,0 +1,43 @@
+
+EVP_PKEY_set_alias_type() allows modifying a EVP_PKEY to use a
+different set of algorithms than the default. This is currently used
+to support SM2 keys, which use an identical encoding to ECDSA.
```
I think adding a "=head1 EXAMPLE" section might be helpful.

I do also wonder whether it would be useful to have an "overview" type man page for SM2, e.g. see the other overviews in doc/man7. Or possibly a combined overview page covering SM2/3/4. Anyway - just a thought, and for a separate PR.

mattcaswell commented on doc/man3/EVP_PKEY_type.md:
```diff
@@ -0,0 +1,43 @@
+
+=head1 COPYRIGHT
+
+Copyright 2006-2018 The OpenSSL Project Authors. All Rights Reserved.
```
Just 2018

mattcaswell commented on test/evp_test.c:
```diff
@@ -2530,6 +2530,11 @@ static int parse(EVP_TEST *t)
+        /* Hack to detect SM2 keys */
+        if(strstr(key->name, "SM2")) {
+            EVP_PKEY_set_alias_type(pkey, EVP_PKEY_SM2);
+        }
```
No "{}" here.

randombit commented:
Believe everything from round2 review is fixed in 7fd81fd. If all is ok let me know and I will rebase+squash the commits.

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    if (sig == NULL) {
+        *siglen = ECDSA_size(ec);
+        return 1;
+    } else if (*siglen < (size_t)ECDSA_size(ec)) {
```
I prefer not to see "else" here.  Not a requirement since I know some poor misguided souls disagree. :)

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+}
+
+static int pkey_sm2_verify(EVP_PKEY_CTX *ctx,
+                          const unsigned char *sig, size_t siglen,
```
Squinting nit: this and next line over one more space?

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+        return 0;
+    }
+
+    if (dctx->md != NULL)
```
Optional: I prefer to see the ?: operator used here.

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    if (dctx->md != NULL)
```
Optional: the ?: operator again.

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx->md != NULL)
```
Optional, you know what I'm gonna say.

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx->md != NULL)
```
...

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
+{
+    int ret;
```
Above you merge ret/type into one declaration.  Consider doing that here.

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
+{
+    int ret = -1;
```
merge?

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+        return 1;
+
+    case EVP_PKEY_CTRL_MD:
+        if (EVP_MD_type((const EVP_MD *)p2) != NID_sha256
```
Introduce a variable to hold EVP_MD_type value and compute once.

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+                            const char *type, const char *value)
+{
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+        int nid;
```
can merge the decl and the init code into one line (but don't forget the blank line)

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+        int nid;
+
+        nid = EC_curve_nist2nid(value);
+        if (nid == NID_undef)
```
I get it, avoiding walking-to-the-right. Perhaps this?
```
if ((nid = EV_curve_nist..) == NID_undef
        && (nid = OBJ_sn2nid...) == NID_undef
        && (nid = OBJ_ln2nid...) == NID_UNDEF) {
    SM2err
...
}
```

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    SM2_PKEY_CTX *dctx = ctx->data;
+    int ret = 0;
+
+    if (dctx->gen_group == NULL) {
```
Is this really along the lines of a NUL check that we would typically avoid?

dot-asm replied:
I'd suggest to remove this check too. As well as second another one...

dot-asm replied:
Just instantiated object would have gen_group set to NULL. One can argue that calling paramgen without setting parameter is programming mistake, but it's probably not severe. So should this should probably stay? @richsalz?

richsalz commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,323 @@
+    EC_KEY *ec = NULL;
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (ctx->pkey == NULL && dctx->gen_group == NULL) {
```
similar, is this an assert-style error which we should coredump on?

dot-asm replied:
As for check for ctx->pkey. Below we observe `if (ctx->pkey != NULL) {} **else** {}`. So which one is it?

As for dctx->gen_group, same as above...

dot-asm replied:
Disregard last remark!!! I mean the if-else is legitimate, because the condition is if *both* are zero.

randombit commented:
All review comments from @richsalz should be addressed by 19c2a45  except for the possibly undesired null pointer checks. I will point out that the EC pmeth has the same checks. Also, while (AIUI) the policy is to not check for NULL pointer arguments, ctx->pkey and dctx->gen_group are internal values set by the library itself. So crashing on a null pointer there is not likely to help an application developer understand that they should have made some previous call to set the parameter in the ctx.

Anyway I can remove them if needed, just let me know what the decision is.

richsalz commented:
Keeping the NULL checks is fine.  I approve this.  Matt can reconfirm or merge as he sees fit. :)

richsalz commented:
Rebase/push-f if you want, don't squash until @mattcaswell has looked at your latest commit.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,306 @@
+    const int md_type = (dctx->md == NULL) ? NID_sm3 : EVP_MD_type(dctx->md);
+
+    if (out == NULL) {
+        if (!sm2_plaintext_size(ec, EVP_get_digestbynid(md_type), inlen,
```
I don't quite follow why does EVP_get_digestbynid have to be invovled here? As simplest case wouldn't `dctx->md ? dctx->md : EVP_sm3()` do the job? If one has to use EVP_get_digestbynid(NID_sm3), then `dctx->md ? dctx->md : EVP_get_digestbynid(NID_sm3). Either way this kind of calls for question if it would be appropriate to check for EVP_get_digestbynid return value. Well, no-sm3 implies no-sm2, so one can argue that lookup for NID_sm3 is bound to succeed [as well as availability of EVPsm3()].

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,306 @@
+
+static int pkey_sm2_encrypt(EVP_PKEY_CTX *ctx,
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
```
Formatting nit, two too many spaces in above two lines.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,306 @@
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+       int nid = NID_undef;
+
+       if(((nid = EC_curve_nist2nid(value)) == NID_undef)
```
Space after `if`.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,306 @@
+
+       if(((nid = EC_curve_nist2nid(value)) == NID_undef)
+          && ((nid = OBJ_sn2nid(value)) == NID_undef)
+          && ((nid == OBJ_ln2nid(value)) == NID_undef)) {
```
First `==` aught to be typo.

randombit commented:
@dot-asm Thanks, your comments should be addressed by 8af1a2c

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,306 @@
+        return 0;
+    }
+
+    ret = sm2_sign(type, tbs, tbslen, sig, &sltmp, ec);
```
`type` makes no sense to me. I mean it's assigned to NID_sm3 or whatever nid dctx->md has. Then it's passed to sm2_sign where it's checked for being NID_sm3 and **not** used in any other way. If sm3 is the only one permitted, I'd say check should be performed earlier...

randombit replied:
I think the final consensus (in discussion in the original SM2 PR) was that SM2 could be used with any "approved" hash. And while it so happens that currently the only approved hash is SM3, SM2 should not make this restriction eg in the event that the Chinese standard is later amended to also accept say SM2 with SHA-256. Also I see GmSSL has OIDs defined for SM2 signatures with SHA-2

So I think the correct action is to remove the check for SM3 and accept signing with any digest.  @ronaldtse any thoughts here?

dot-asm replied:
Checking for type in function that accepts already generated value is not right spot. Check should be done at attempt to generated the value. There are more inconsistencies in code. On one hand it's asserted that it can be used only with given hash algorithm and given curve, but at the same time ctrl accepts practically any curve, multiple digests,,,

randombit replied:
OK I have pushed f481ccfe304645 which removes the restriction on using SM3 or a 256-bit digest with SM2 signatures. I generated a test signature using GmSSL with SHA-512 hash and added it to the evp tests.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,296 @@
+            && md_nid != NID_sha512
+            && md_nid != NID_sm3) {
+            SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_DIGEST_TYPE);
+            return 0;
```
Wouldn't it be appropriate to reject all but sm3 here? If not, then question would also be following, Check for nid in e.g. sm2_sign reads `if (type != NID_sm3 || dgstlen != 32)`, so how does latter check for length align with say sha512 being permitted here? Then it's not like you enforce sm2 curve, so another curve would be able to handle it? But then check for length it not right... Kind of a mess...

dot-asm commented on crypto/evp/pmeth_lib.c:
```diff
@@ -104,7 +107,8 @@ static EVP_PKEY_CTX *int_ctx_new(EVP_PKEY *pkey, ENGINE *e, int id)
             return NULL;
-        id = pkey->ameth->pkey_id;
+
+        id = pkey->type;
```
While you are at it, could you replace first logical `!` with explicit NULL check and omit second one. If also appropriate to add empty line  prior outermost if... Actually one can wonder if one needs the NULL check at all... I mean if subroutine ends up being called with pkey being NULL and id being -1, then can we view it as programmer's mistake (or memory corruption)? Because if we can, then it would be appropriate to omit the check.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,283 @@
+
+static int pkey_sm2_decrypt(EVP_PKEY_CTX *ctx,
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
```
Formatting nit, two too many spaces in above two lines.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,283 @@
+    case EVP_PKEY_CTRL_MD: {
+        dctx->md = p2;
+        return 1;
+    }
```
Omit {}.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,283 @@
+}
+
+static int pkey_sm2_ctrl_str(EVP_PKEY_CTX *ctx,
+                            const char *type, const char *value)
```
Formatting nit, one space short in above line.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,283 @@
+                            const char *type, const char *value)
+{
+    if (strcmp(type, "ec_paramgen_curve") == 0) {
+       int nid = NID_undef;
```
Formatting nit, one space short in above line. As well as in line below it. And following two lines are two spaces short...

dot-asm commented:
I'd like to suggest to remaster this request into two commits. One that adds EVP_PKEY_set_alias_type (code and doc), and modifies int_ctx_new, and second commit that does the rest.

randombit commented:
@dot-asm Splitting into 2 commits as you suggest seems good to me. I will do this once I have a final all-clear on the changes.

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,282 @@
+}
+
+static int pkey_sm2_sign(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
+                        const unsigned char *tbs, size_t tbslen)
```
Another formatting nit, one space short.

dot-asm commented:
> There were two suggestions to remove NULL-checks.

I'm having another look. There might be some more remarks, so hold it for now...

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,282 @@
+    dctx = dst->data;
+    if (sctx->gen_group != NULL) {
+        dctx->gen_group = EC_GROUP_dup(sctx->gen_group);
+        if (dctx->gen_group != NULL)
```
This should be check for == NULL. Secondly since error returned it's only appropriate to free whatever was allocated in pkey_sm2_init. In other words
```
if (dctx->gen_group == NULL) {
    pkey_sm2_cleanup(dst);
    return 0;
}
```
One can also wonder if it would be appropriate to flip the order and first attempt EC_GROUP_dup prior even calling pkey_sm2_init.
```

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,282 @@
+        return 1;
+    }
+
+    if (*siglen < (size_t)ECDSA_size(ec)) {
```
**Formally speaking** this is problematic cast.  Because if ECDSA_size returns negative value, then this condition would be met. Well, same can be said about above call to ECDSA_size. Two options, a) leave comment in the beginning that we count on ECDSA_size to return small positive (one can express it in form of "soft" assert); b) explicitly check for ECDSA_size return value being positive in the beginning of function and return failure if it's not. In either case it would be appropriate to add local variable that accepts ECDSA_size value.

randombit replied:
I will fix here, same logic is used in ec_pmeth.c btw

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,282 @@
+        if (!EC_KEY_set_group(ec, dctx->gen_group))
+            return 0;
+    }
+    return EC_KEY_generate_key(pkey->pkey.ec);
```
This is unnecessary hard to read. I mean where does pkey->pkey.ec come into picture? Yes, it's assigned above EVP_PKEY_assign_EC_KEY, but it's unnecessary hard to make the connection. Besides one can also wonder if following pattern would be more appropriate to
```
ec = EC_KEY_new();
...
if (EC_KEY_generate_key(ec))
    EVP_PKEY_assign_EC_KEY(pkey,ec);
else
    ECP_KEY_free();
```
It looks like it was copy-n-pasted from from ec_pmeth.c, so that one can say all same things about ec_pmeth.c...

dot-asm replied:
> It looks like it was copy-n-pasted from from ec_pmeth.c, so that one can say all same things about ec_pmeth.c...

Which means that it would be appropriate to harmonize eevn ec_pmeth.c afterwards...

randombit replied:
Hopefully better now

dot-asm commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    return EVP_PKEY_assign_EC_KEY(pkey, ec);
+}
```
I'd like to suggest "ultimate" solution, **omit** pkey_sm2_paramgen and pkey_sm2_keygen. There is no any SM2 "twist" to keygen and default one from common pmeth would work [with the curve in question]. Application would even have to do anything special. It should also be noted that these two are not exercised in any way, so we don't actually know if it works...

dot-asm replied:
> Application would even have to do anything special.

I meant to write "application would **not** even have to do anything special."

dot-asm replied:
Just to clarify, there actually are more problems with paramgen and keygen. But instead of addressing the problems suggestion is to omit the functions in question from sm2_pmeth. This doesn't affect user's ability to generate keys, because it can be done with "standard" ec_pmeth. If at later point it turns out that sm2_pmeth has to have these two, then they can be added then. Preferably as "bypass" to ec_pmeth. By "bypass" I mean that since sm2_pmeth is an "alias" pmeth, it would be appropriate to engage "parent" pmeth for functions that are not performing method-specific operations.

Just in case. The fact that there are more problems with paramgen and keygen actually means that there are problems even in ec_pmeth. I plan to address them as soon as we are ready with this one.

dot-asm commented:
Speaking of dead code, why aren't there Sign and Encrypt tests?

randombit commented:
Sign and encrypt are tested by `sm2_internal_test.c`. They are not tested by `evp_test` for the same reason ECDSA signing is not tested there, it does not seem possible to set the nonce.

mattcaswell commented:
> Sign and encrypt are tested by sm2_internal_test.c. They are not tested by evp_test for the same reason ECDSA signing is not tested there, it does not seem possible to set the nonce.

Perhaps something in evp_extra_test.c might be appropriate? i.e. just sign something, and then check that the result verifies. It doesn't prove that signing produces *correct* output as per the standard, but it at least checks that a sign/verify roundtrip produces sane results and doesn't unexpectedly fail somewhere. Similarly with encrypt/decrypt.

randombit commented:
@mattcaswell @dot-asm I added a roundtrip test for signatures and encryption as suggested, and removed the SM2 specific pmeth for paramgen and keygen.

randombit commented:
CI build is now failing with this error

> EVP function code 204 collision at EVP_F_EVP_PKEY_SET_ALIAS_TYPE

After rebasing against master I see the same problem locally. It appears another EVP function was added recently which caused a conflicting assignment. How do I resolve this?

dot-asm commented:
Looks like time to remaster as two commits, as suggested before.

randombit commented:
Rebased against master and converted into two commits, one for EVP_set_alias_type and other for SM2 specific changes.

randombit commented:
Memory leaks in new EVP tests fixed, and I also got the SM2 errors out of the first commit.

dot-asm commented:
Cool, references to deleted functions are gone from sm2_err.c. :-)

romen commented:
So IIUC, once this PR gets merged into master, it will be possible from a program to explicitly use `EVP_PKEY_set_alias()` to force an existing EC `EVP_PKEY` to use the SM2 `PKEY_METH`, basically making it possible for a purposefully made program to perform SM2DSA (or encryption) over any EC curve, including the recommended SM2 curve. Is this correct?

But, how can a *generic* application reading a PEM file containing the private key decide to use SM2DSA over ECDSA?
For example: when using the `apps/digest.c` app to sign something, IIUC, after this PR upon reading an EC private key on the SM2 recommended curve from the specified PEM file will actually do ECDSA over that curve, rather than SM2DSA.
Other example: `test/evp_test.c`, lines 2534+ force keys over the SM2 recommended curve to use the SM2 pmeth using `EVP_PKEY_set_alias()`. As a result it is not possible to write evp_tests to test ECDSA over the SM2 recommended curve, nor to test SM2DSA signatures over NIST-P256 or any other curve (excluding the SM2 recommended curve).

Additionally, SM2DSA requires the userid/ZA/personalization field (e.g., "alice@yahoo.com") to generate a signature, but I don't see a way to provide such input through the `EVP` API, and this is further exemplified by the fact that none of the tests using the `EVP` API  (not in `test/recipes/30-test_evp_data/evppkey.txt` nor in `test/evp_extra_test.c`) is feeding it as an input value.

randombit commented:
> But, how can a generic application reading a PEM file containing the private key decide to use SM2DSA over ECDSA?

It cannot, because SM2 uses the same key format as ECDSA and ECDH (and ECIES, ECGDSA, ECKCDSA, ...) and there is no way from looking at the bits to tell which algorithm is intended beyond "an EC algorithm of some kind".

IMO that's fine because a generic application probably should not be using off-brand EC signatures without a very conscious decision to support such things on the part of the developer.

> As a result it is not possible to write evp_tests to test ECDSA over the SM2 recommended curve, nor to test SM2DSA signatures over NIST-P256 or any other curve (excluding the SM2 recommended curve).

This is not quite right - in the EVP tests if the key is given a name including substring "SM2" then SM2 is used. It has nothing to do with the key params. So there is no problem testing ECDSA with SM2 curve, or SM2 with P-256. (On current master it is true that SM2 the algo and SM2 the curve are tied together in away that prevents the usages you suggest.)

> Additionally, SM2DSA requires the userid/ZA/personalization field

Indeed. I think this functionally got lost when the SM2 API functions were made private. However it seems GmSSL also allows creating signatures without the field (demonstrable by the fact that the EVP test data, which I generated with GmSSL command line util, does not use the personalization functionality). But, end users will likely want this. Probably this could be exposed via an EVP_PKEY_ctrl.

In the interests of making forward progress I would prefer having this PR (addressing #5924) merged and then come back with another PR adding a ctrl for setting the personalization string. But of course I defer to openssl team on that call.

mattcaswell commented:
> > But, how can a generic application reading a PEM file containing the private key decide to use SM2DSA over ECDSA?
>
> It cannot, because SM2 uses the same key format as ECDSA and ECDH (and ECIES, ECGDSA, ECKCDSA, ...) and there is no way from looking at the bits to tell which algorithm is intended beyond "an EC algorithm of some kind".
>
> IMO that's fine because a generic application probably should not be using off-brand EC signatures without a very conscious decision to support such things on the part of the developer.

I agree. We could imagine some command line flag which signals that a different alias should be used ("-alias" perhaps?). But certainly not for this PR.

> In the interests of making forward progress I would prefer having this PR (addressing #5924) merged and then come back with another PR adding a ctrl for setting the personalization string. But of course I defer to openssl team on that call.

I agree.

romen commented:
Thanks @randombit for clarifying!

@mattcaswell yes, I too think it makes sense, I just wanted to make sure I correctly understood the current limits/caveats of the SM2 pmeth, and raise a flag in case some of them required to be addressed before merging!

mattcaswell commented on crypto/sm2/sm2_pmeth.c:
```diff
@@ -0,0 +1,245 @@
+    sctx = src->data;
+    dctx = dst->data;
+    if (sctx->gen_group != NULL) {
+        dctx->gen_group = EC_GROUP_dup(sctx->gen_group);
```
Shouldn't we free any pre-existing `dctx->gen_group` first?

dot-asm replied:
dctx->gen_group is guaranteed to be NULL after pkey_sm2_init. This however ask for question if pkey_sm2_init should see if there ctx->data is not NULL. pkey_ec_init in ec_meth.c doesn't, so at least pkey_sm2_init consistent with that.

mattcaswell commented:
Pushed!! Thanks @randombit! Will you follow up with a PR to handle the personalisation string? We should also consider whether we want any changes to the command line apps

randombit commented:
@mattcaswell No problem, probably have something by end of this week for setting personalization string.

ronaldtse commented:
Thank you @randombit @mattcaswell ! 👍
